api/v1alpha1/clustertemplateinstance_utils.go

```diff
@@ -292,3 +293,100 @@ func (i *ClusterTemplateInstance) GetHelmParameters(
 
 	return params, nil
 }
+
+func (i *ClusterTemplateInstance) GetSubjectsWithClusterTemplateUserRole(ctx context.Context, k8sClient client.Client) (subjects []rbacv1.Subject, err error) {
+	allRoleBindingsInNamespace := &rbacv1.RoleBindingList{}
+
+	if err := k8sClient.List(ctx, allRoleBindingsInNamespace, &client.ListOptions{
+		Namespace: i.Namespace,
+	}); err != nil {
+		return nil, err
+	}
+
+	result := []rbacv1.Subject{}
+	keys := make(map[string]bool)
+	for _, rb := range allRoleBindingsInNamespace.Items {
+		if rb.RoleRef.Kind == "ClusterRole" && rb.RoleRef.Name == "cluster-templates-user" {
+			for _, subject := range rb.Subjects {
+				key := subject.Kind + "*" + subject.Name
+				if _, value := keys[key]; !value {
+					keys[key] = true
+					result = append(result, subject)
+				}
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func (i *ClusterTemplateInstance) CreateDynamicRole(ctx context.Context, k8sClient client.Client, secretNames []string) (r *rbacv1.Role, err error) {
+	roleName := i.Name + "-role-managed"
+	roleNamespace := i.Namespace
+
+	existingRole := &rbacv1.Role{}
+	err = k8sClient.Get(
+		ctx,
+		client.ObjectKey{
+			Name:      roleName,
+			Namespace: roleNamespace,
+		},
+		existingRole,
+	)
+
+	desiredRole := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            roleName,
+			Namespace:       roleNamespace,
+			OwnerReferences: []metav1.OwnerReference{i.GetOwnerReference()},
+		},
+		Rules: []rbacv1.PolicyRule{{
+			APIGroups:     []string{""},
+			Verbs:         []string{"Get"},
+			Resources:     []string{"secrets"},
+			ResourceNames: secretNames,
+		}},
+	}
+
+	if err == nil {
+		// Results in no action if there is no difference in content
+		return desiredRole, k8sClient.Update(ctx, desiredRole)
+	} else {
+		return desiredRole, k8sClient.Create(ctx, desiredRole)
+	}
+}
+
+func (i *ClusterTemplateInstance) CreateDynamicRoleBinding(ctx context.Context, k8sClient client.Client, role *rbacv1.Role, roleSubjects []rbacv1.Subject) (rb *rbacv1.RoleBinding, err error) {
```

Comment:
Lets remove the named import too
```suggestion
func (i *ClusterTemplateInstance) CreateDynamicRoleBinding(ctx context.Context, k8sClient client.Client, role *rbacv1.Role, roleSubjects []rbacv1.Subject) (*rbacv1.RoleBinding, error) {
```